src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/OptionDescriptorExtensions.cs

```diff
@@ -0,0 +1,89 @@
+            IEnumerable<IOptionDescriptor<object>> descriptors)
+        {
+            var instance = descriptors.Single(d => d.OptionType == typeof(TInstance)).Instance;
+            return (TInstance)instance;
+        }
+
+        /// <summary>
+        /// Returns the only instance of <typeparamref name="TInstance"/> from a sequence of option descriptors, 
+        /// or a default value if the sequence is empty.
+        /// </summary>
+        /// <typeparam name="TInstance">The type of the instance described by the option descriptor.</typeparam>
+        /// <param name="descriptors">A sequence of <see cref="IOptionDescriptor{object}"/>.</param>
+        /// <returns>The only instance of <typeparamref name="TInstance"/> from a sequence of 
+        /// <see cref="IOptionDescriptor{object}"/>, 
+        /// or a default value if the sequence is empty; throws an exception if there is more than one 
+        /// <typeparamref name="TInstance"/> 
+        /// in the sequence.</returns>
+        public static TInstance InstanceOfOrDefault<TInstance>([NotNull] this 
+            IEnumerable<IOptionDescriptor<object>> descriptors)
+        {
+            TInstance instance = default(TInstance);
+            var descriptor = descriptors.SingleOrDefault(d => d.OptionType == typeof(TInstance));
+            if (descriptor != null)
+            {
+                instance = (TInstance)descriptor.Instance;
+            }
+            return instance;
+        }
+
+        /// <summary>
+        /// Returns all the instances of <typeparamref name="TInstance"/> from a sequence of option descriptors.
+        /// </summary>
+        /// <typeparam name="TInstance">The type of the instance described by the option descriptor.</typeparam>
+        /// <param name="descriptors">A sequence of <see cref="IOptionDescriptor{object}"/>.</param>
+        /// <returns>An IEnumerable of <typeparamref name="TInstance"/> that contains instances from a sequence 
+        /// of option descriptors.</returns>
+        public static IEnumerable<TInstance> InstancesOf<TInstance>([NotNull] this 
+            IEnumerable<IOptionDescriptor<object>> descriptors)
+        {
+            var instances = descriptors
+                .Where(d => d.OptionType == typeof(TInstance))
+                .Select(d => d.Instance);
+            return instances.Cast<TInstance>();
+        }
+
+        /// <summary>
+        /// Removes all instances of <typeparamref name="TInstance"/> from a list of option descriptors.
+        /// </summary>
+        /// <typeparam name="TInstance">The type of the instance described by the option descriptor.</typeparam>
+        /// <typeparam name="TOptionDescriptor">A type that implements 
+        /// <see cref="IOptionDescriptor{object}"/>.</typeparam>
+        /// <param name="descriptors">A sequence of <see cref="IOptionDescriptor{object}"/>.</param>
+        public static void RemoveTypesOf<TInstance, TOptionDescriptor>([NotNull] this 
```

Comment:
Casting from IEnumerable to IList won't work.  That leaves explicit extension methods for `RemoveTypesOf`, and possibly `ReplaceTypesOf`, on `IList<InputFormatterDescriptor>` and `IList<OutputFormatterDescriptor>`.  Would you like me to do that?

Just to be sure, `RemoveTypesOf` should also remove descriptors with null Instances?  That's the way it's written now.